ext/ScreenShot/ScreenShot.h

```diff
@@ -5,27 +5,151 @@
+
+					auto imageView = asset::ICPUImageView::create(std::move(imgViewInfo));
+					auto gpuImageView = driver->getGPUObjectsFromAssets(&imageView.get(), &imageView.get() + 1)->front();
+
+					return std::move(gpuImageView);
+				};
+
+				auto gpuImageViewDepthBuffer = createAttachement(false);
+				auto gpuImageViewColorBuffer = createAttachement(true);
+
+				auto frameBuffer = driver->addFrameBuffer();
+				frameBuffer->attach(video::EFAP_DEPTH_ATTACHMENT, std::move(gpuImageViewDepthBuffer));
+				frameBuffer->attach(video::EFAP_COLOR_ATTACHMENT0, std::move(gpuImageViewColorBuffer));
+
+				return frameBuffer;
+			};
+
+			/*
+				Create a ScreenShot with gpu image usage.
+			*/
+
+			bool createScreenShoot(core::smart_refctd_ptr<IrrlichtDevice> device, core::smart_refctd_ptr<video::IGPUImage> gpuImage, const std::string& outFileName)
+			{
+				auto driver = device->getVideoDriver();
+				auto assetManager = device->getAssetManager();
+
+				auto fetchedImageParams = gpuImage->getCreationParameters();
+				auto image = asset::ICPUImage::create(std::move(fetchedImageParams));
+
+				auto trueBufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(fetchedImageParams.extent.width, asset::getTexelOrBlockBytesize(fetchedImageParams.format));
+
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+				asset::ICPUImage::SBufferCopy& region = regions->front();
+
+				region.imageSubresource.mipLevel = 0u;
+				region.imageSubresource.baseArrayLayer = 0u;
+				region.imageSubresource.layerCount = 1u;
+				region.bufferOffset = 0u;
+				region.bufferRowLength = trueBufferRowLength;
+				region.bufferImageHeight = 0u;
+				region.imageOffset = { 0u, 0u, 0u };
+				region.imageExtent = image->getCreationParameters().extent;
+
+				auto destinationBuffer = core::smart_refctd_ptr<video::IGPUBuffer>(driver->createDownStreamingGPUBufferOnDedMem(image->getImageDataSizeInBytes()));
+				destinationBuffer->getBoundMemory()->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ, { 0u, destinationBuffer->getSize() });
+
+				driver->copyImageToBuffer(gpuImage.get(), destinationBuffer.get(), 1, regions->begin());
+
+				auto texelBuffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(image->getImageDataSizeInBytes());
+				auto rawData = reinterpret_cast<uint8_t*>(destinationBuffer->getBoundMemory()->getMappedPointer());
+
+				memcpy(texelBuffer->getPointer(), rawData, image->getImageDataSizeInBytes());
+
+				image->setBufferAndRegions(std::move(texelBuffer), regions);
+
+				asset::IAssetWriter::SAssetWriteParams wparams(image.get());
+				return assetManager->writeAsset(outFileName, wparams);
+			}
+
+			//! TODO: HANDLE UNPACK ALIGNMENT
+			core::smart_refctd_ptr<video::IDriverFence> createScreenShot(video::IDriver* driver, video::IGPUImage* source, video::IGPUBuffer* destination, uint32_t sourceMipLevel = 0u, size_t destOffset = 0ull, bool implicitflush = true)
```

Comment:
this should probably be called `downloadImageMipLevel`

Reply:
actually I don't know what exactly fence is, need to read about it

Reply:
Vulkan Event/Semaphore, OpenGL GLsync